open_spiel/python/mfg/algorithms/distribution.py

```diff
@@ -68,33 +68,34 @@ def evaluate(self):
 
     while not self.is_terminal_from_states(listing_states):
       new_listing_states = []
+      new_distribution = collections.defaultdict(float)
 
       if self.type_from_states(listing_states) == pyspiel.StateType.CHANCE:
         for mfg_state in listing_states:
           for action, prob in mfg_state.chance_outcomes():
             new_mfg_state = mfg_state.child(action)
             new_mfg_state_str = new_mfg_state.observation_string(
                 pyspiel.PlayerId.DEFAULT_PLAYER_ID)
-            if new_mfg_state_str not in self.distribution:
+            # As a state can be the child of two different parent states, we do
+            # not add it in the new states if it has already been seen in this
+            # iteration.
+            if new_mfg_state_str not in new_distribution:
               new_listing_states.append(new_mfg_state)
-            self.distribution[new_mfg_state_str] += prob * self.distribution[
-                mfg_state.observation_string(
-                    pyspiel.PlayerId.DEFAULT_PLAYER_ID)]
+            new_distribution[new_mfg_state_str] += prob * self.value(mfg_state)
 
       elif (self.type_from_states(listing_states) ==
             pyspiel.StateType.MEAN_FIELD):
         for mfg_state in listing_states:
           dist_to_register = mfg_state.distribution_support()
-          dist = [
-              self.distribution[str_state] for str_state in dist_to_register
-          ]
+          dist = list(map(self.value_str, dist_to_register))
+          assert (sum(dist) - len(self._root_states)) < 1e-4, (
```

Comment:
self._root_states -> self.game.num_players() might be clearer.

Reply:
Done. I had to add the game as an attribute of the distribution.